spec/support/mock_intake.rb

```diff
@@ -187,3 +187,116 @@ def print_received
 
   config.include MockIntake::WaitFor, :mock_intake
 end
+
+class EventCollector
+  class << self
+    def method_missing(name, *args, &block)
+      instance.send(name, *args, &block)
+    end
+
+    def instance
+      @instance ||= new
+    end
+  end
+
+  attr_reader(
+      :errors,
+      :metadatas,
+      :metricsets,
+      :requests,
+      :spans,
+      :transactions
+  )
+
+  def initialize
+    clear!
+  end
+
+  def catalog(json)
+    case json.keys.first
+    when 'transaction' then transactions << json.values.first
+    when 'span' then spans << json.values.first
+    when 'error' then errors << json.values.first
+    when 'metricset' then metricsets << json.values.first
+    when 'metadata' then metadatas << json.values.first
+    end
+  end
+
+  def clear!
+    @requests = []
+
+    @errors = []
+    @metadatas = []
+    @metricsets = []
+    @spans = []
+    @transactions = []
+  end
+
+  def transaction_metrics
+    metricsets.select { |set| set['transaction'] && !set['span'] }
+  end
+
+  def span_metrics
+    metricsets.select { |set| set['transaction'] && set['span'] }
+  end
```

Comment:
I'm not sure I think these are generic enough to extract.

Reply:
Do you mean as opposed to using the current inline block approach (i.e. [this strategy](https://github.com/elastic/apm-agent-ruby/blob/master/spec/integration/rails_spec.rb#L294-L296))? I got a lot of random test failures with jRuby using the block approach, where the `EventCollector`/`MockIntake` object was `nil` inside the block . I therefore decided to do a more explicit approach using a method and those random failures went away. I can't explain why but I'm guess it has something to do with defining a block's binding in JRuby.

Reply:
i.e. this works so I'd rather have an explicit method even though it's not perfect object interface design.

Reply:
We can keep them as is for now if you'd like 👍

Reply:
I'm afraid to change anything 😬

Reply:
Welcome to the `rails_spec.rb`